File: bugsnag-android-performance-impl/src/testFixtures/java/com/bugsnag/android/performance/test/TestTimeoutExecutor.java
package com.bugsnag.android.performance.test;

import com.bugsnag.android.performance.internal.processing.Timeout;
import com.bugsnag.android.performance.internal.processing.TimeoutExecutor;

import org.jetbrains.annotations.NotNull;

import java.util.PriorityQueue;

public class TestTimeoutExecutor implements TimeoutExecutor {
    private final PriorityQueue<Timeout> timeouts = new PriorityQueue<>();

    public void runAllTimeouts() {
        while (!timeouts.isEmpty()) {
            Timeout timeout = timeouts.poll();
            if (timeout != null) {
                timeout.run();
            }
        }
    }

    @Override
    public void scheduleTimeout(@NotNull Timeout timeout) {
        timeouts.add(timeout);
    }

    @Override
    public void cancelTimeout(@NotNull Timeout timeout) {
        timeouts.remove(timeout);
    }
}

File: bugsnag-android-performance/src/test/java/com/bugsnag/android/performance/test/TestTimeoutExecutor.kt
package com.bugsnag.android.performance.test

import com.bugsnag.android.performance.internal.processing.Timeout
import com.bugsnag.android.performance.internal.processing.TimeoutExecutor
import java.util.PriorityQueue

internal class TestTimeoutExecutor : TimeoutExecutor {
    internal val timeouts = PriorityQueue<Timeout>()

    fun runAllTimeouts() {
        var timeout = timeouts.poll()
        while (timeout != null) {
            timeout.run()
            timeout = timeouts.poll()
        }
    }

    override fun scheduleTimeout(timeout: Timeout) {
        timeouts.add(timeout)
    }

    override fun cancelTimeout(timeout: Timeout) {
        timeouts.remove(timeout)
    }
}
